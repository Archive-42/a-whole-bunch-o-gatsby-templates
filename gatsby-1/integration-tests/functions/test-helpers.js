const fetch = require(`node-fetch`)
const { createReadStream } = require("fs")
const execa = require(`execa`)
const fs = require(`fs-extra`)
const path = require(`path`)
const FormData = require("form-data")

export function runTests(env, host) {
  describe(env, () => {
    describe(`routing`, () => {
      test(`top-level API`, async () => {
        const result = await fetch(`${host}/api/top-level`).then(res =>
          res.text()
        )

        expect(result).toMatchSnapshot()
      })
      test(`secondary-level API`, async () => {
        const result = await fetch(
          `${host}/api/a-directory/function`
        ).then(res => res.text())

        expect(result).toMatchSnapshot()
      })
      test(`secondary-level API with index.js`, async () => {
        const result = await fetch(`${host}/api/a-directory`).then(res =>
          res.text()
        )

        expect(result).toMatchSnapshot()
      })
      test(`secondary-level API`, async () => {
        const result = await fetch(`${host}/api/dir/function`).then(res =>
          res.text()
        )

        expect(result).toMatchSnapshot()
      })
      test(`routes with special characters`, async () => {
        const routes = [
          `${host}/api/I-Am-Capitalized`,
          `${host}/api/some whitespace`,
          `${host}/api/with-äöü-umlaut`,
          `${host}/api/some-àè-french`,
          encodeURI(`${host}/api/some-אודות`),
        ]

        for (const route of routes) {
          const result = await fetch(route).then(res => res.text())

          expect(result).toMatchSnapshot()
        }
      })

      test(`dynamic routes`, async () => {
        const routes = [
          `${host}/api/users/23/additional`,
          `${host}/api/dir/super`,
        ]

        for (const route of routes) {
          const result = await fetch(route).then(res => res.json())

          expect(result).toMatchSnapshot()
        }
      })
    })

    describe(`environment variables`, () => {
      test(`can use inside functions`, async () => {
        const result = await fetch(`${host}/api/env-variables`).then(res =>
          res.text()
        )

        expect(result).toEqual(`word`)
      })
    })

    describe(`typescript`, () => {
      test(`typescript functions work`, async () => {
        const result = await fetch(`${host}/api/i-am-typescript`).then(res =>
          res.text()
        )

        expect(result).toMatchSnapshot()
      })
    })

    describe(`function errors don't crash the server`, () => {
      // This test mainly just shows that the server doesn't crash.
      test(`normal`, async () => {
        const result = await fetch(`${host}/api/error-send-function-twice`)

        expect(result.status).toEqual(200)
      })
    })

    describe(`response formats`, () => {
      test(`returns json correctly`, async () => {
        const res = await fetch(`${host}/api/i-am-json`)
        const result = await res.json()

        const { date, ...headers } = Object.fromEntries(res.headers)
        expect(result).toMatchSnapshot()
        expect(headers).toMatchSnapshot()
      })
      test(`returns text correctly`, async () => {
        const res = await fetch(`${host}/api/i-am-typescript`)
        const result = await res.text()

        const { date, ...headers } = Object.fromEntries(res.headers)
        expect(result).toMatchSnapshot()
        expect(headers).toMatchSnapshot()
      })
    })

    describe(`functions can send custom statuses`, () => {
      test(`can return 200 status`, async () => {
        const res = await fetch(`${host}/api/status`)

        expect(res.status).toEqual(200)
      })

      test(`can return 404 status`, async () => {
        const res = await fetch(`${host}/api/status?code=404`)

        expect(res.status).toEqual(404)
      })

      test(`can return 500 status`, async () => {
        const res = await fetch(`${host}/api/status?code=500`)

        expect(res.status).toEqual(500)
      })
    })

    describe(`functions can parse different ways of sending data`, () => {
      test(`query string`, async () => {
        const result = await fetch(
          `${host}/api/parser?amIReal=true`
        ).then(res => res.json())

        expect(result).toMatchSnapshot()
      })

      test(`form parameters`, async () => {
        const { URLSearchParams } = require("url")
        const params = new URLSearchParams()
        params.append("a", `form parameters`)
        const result = await fetch(`${host}/api/parser`, {
          method: `POST`,
          body: params,
        }).then(res => res.json())

        expect(result).toMatchSnapshot()
      })

      test(`form data`, async () => {
        const FormData = require("form-data")

        const form = new FormData()
        form.append("a", `form-data`)
        const result = await fetch(`${host}/api/parser`, {
          method: `POST`,
          body: form,
        }).then(res => res.json())

        expect(result).toMatchSnapshot()
      })

      test(`json body`, async () => {
        const body = { a: `json` }
        const result = await fetch(`${host}/api/parser`, {
          method: `POST`,
          body: JSON.stringify(body),
          headers: { "Content-Type": "application/json" },
        }).then(res => res.json())

        expect(result).toMatchSnapshot()
      })

      test(`file in multipart/form`, async () => {
        const file = createReadStream(
          path.join(__dirname, "./__tests__/fixtures/test.txt")
        )

        const form = new FormData()
        form.append("file", file)
        const result = await fetch(`${host}/api/parser`, {
          method: `POST`,
          body: form,
        }).then(res => res.json())

        expect(result).toMatchSnapshot()
      })

      // TODO enable when functions support streaming files.
      // test(`stream a file`, async () => {
      // const { createReadStream } = require("fs")

      // const stream = createReadStream(path.join(__dirname, "./fixtures/test.txt"))
      // const res = await fetch(`${host}/api/parser`, {
      // method: `POST`,
      // body: stream,
      // })

      // console.log(res)

      // expect(result).toMatchSnapshot()
      // })
    })

    describe(`functions get parsed cookies`, () => {
      test(`cookie`, async () => {
        const result = await fetch(`${host}/api/cookie-me`, {
          headers: { cookie: `foo=blue;` },
        }).then(res => res.json())

        expect(result).toMatchSnapshot()
      })
    })

    describe(`functions can redirect`, () => {
      test(`normal`, async () => {
        const result = await fetch(`${host}/api/redirect-me`)

        expect(result.url).toEqual(host + `/`)
      })
    })

    describe(`functions can have custom middleware`, () => {
      test(`normal`, async () => {
        const result = await fetch(`${host}/api/cors`)

        const headers = Object.fromEntries(result.headers)
        expect(headers[`access-control-allow-origin`]).toEqual(`*`)
      })
    })

    describe(`plugins can declare functions and they can be shadowed`, () => {
      test(`shadowing`, async () => {
        const result = await fetch(
          `${host}/api/gatsby-plugin-cool/shadowed`
        ).then(res => res.text())
        expect(result).toEqual(`I am shadowed`)

        const result2 = await fetch(
          `${host}/api/gatsby-plugin-cool/not-shadowed`
        ).then(res => res.text())
        expect(result2).toEqual(`I am not shadowed`)
      })
      test(`plugins can't declare functions outside of their namespace`, async () => {
        const result = await fetch(
          `${host}/api/i-will-not-work-cause-namespacing`
        )
        expect(result.status).toEqual(404)
      })
    })

    describe(`typescript files are resolved without needing to specify their extension`, () => {
      test(`typescript`, async () => {
        const result = await fetch(`${host}/api/extensions`).then(res =>
          res.text()
        )
        expect(result).toEqual(`hi`)
      })
    })

    // TODO figure out why this gets into endless loops
    // describe.only(`hot reloading`, () => {
    // const fixturesDir = path.join(__dirname, `fixtures`)
    // const apiDir = path.join(__dirname, `../src/api`)
    // beforeAll(() => {
    // try {
    // fs.unlinkSync(path.join(apiDir, `function-a.js`))
    // } catch (e) {
    // // Ignore as this should mostly error with file not found.
    // // We delete to be sure it's not there.
    // }
    // })
    // afterAll(() => {
    // fs.unlinkSync(path.join(apiDir, `function-a.js`))
    // })

    // test(`new function`, cb => {
    // fs.copySync(
    // path.join(fixturesDir, `function-a.js`),
    // path.join(apiDir, `function-a.js`)
    // )
    // setTimeout(async () => {
    // const result = await fetch(
    // `${host}/api/function-a`
    // ).then(res => res.text())

    // console.log(result)
    // expect(result).toMatchSnapshot()
    // cb()
    // }, 400)
    // })
    // })
  })
}
