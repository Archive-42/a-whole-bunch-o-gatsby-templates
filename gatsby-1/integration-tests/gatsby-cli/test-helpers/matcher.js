export const createLogsMatcher = output => {
  return {
    // Useful for debugging
    logOutput() {
      console.log(output)
    },

    should: {
      contain: match => {
        // ink will auto wrap things, so we need to get rid of any whitespace specific checks
        // and let it just make sure there is whitespace
        expect(output).toMatch(new RegExp(match.replace(/\s+/g, `\\s+`)))
      },
    },
  }
}
