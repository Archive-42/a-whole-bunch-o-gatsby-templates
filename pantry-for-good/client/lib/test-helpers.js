import React from 'react'

class Fields extends Array {
  remove = index => super.splice(index, 1)
  push = item => super.push(item || {})
}

export function getMockField({meta, input} = {}) {
  return function MockField(props) {
    const {component: Component, ...rest} = props
    return <Component meta={meta || {}} input={input} {...rest} />
  }
}

export function getFieldArrayMock(fields = []) {
  return function MockFieldArray(props) {
    const {component: Component, ...rest} = props
    return (
      <Component
        fields={new Fields(...fields)}
        {...rest}
      />
    )
  }
}
