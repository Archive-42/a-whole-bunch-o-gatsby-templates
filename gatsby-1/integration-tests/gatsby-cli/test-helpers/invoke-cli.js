import execa, { sync } from "execa"
import { join } from "path"
import strip from "strip-ansi"
import { createLogsMatcher } from "./matcher"

const gatsbyBinLocation = join(
  __dirname,
  "..",
  `node_modules`,
  `gatsby-cli`,
  `cli.js`
)

// Use as `GatsbyCLI.cwd('execution-folder').invoke('new', 'foo')`
export const GatsbyCLI = {
  from(relativeCwd) {
    return {
      invoke(args) {
        const NODE_ENV = args[0] === `develop` ? `development` : `production`
        try {
          const results = sync("node", [gatsbyBinLocation].concat(args), {
            cwd: join(__dirname, `../`, `./${relativeCwd}`),
            env: { NODE_ENV, CI: 1, GATSBY_LOGGER: `ink` },
          })

          return [
            results.exitCode,
            createLogsMatcher(strip(results.stdout.toString())),
          ]
        } catch (err) {
          return [
            err.exitCode,
            createLogsMatcher(strip(err.stdout?.toString() || ``)),
          ]
        }
      },

      invokeAsync: (args, onExit) => {
        const NODE_ENV = args[0] === `develop` ? `development` : `production`
        const res = execa("node", [gatsbyBinLocation].concat(args), {
          cwd: join(__dirname, `../`, `./${relativeCwd}`),
          env: { NODE_ENV, CI: 1, GATSBY_LOGGER: `ink` },
        })

        let logs = ``
        res.stdout.on("data", data => {
          if (!res.killed) {
            logs += data.toString()
          }

          if (onExit && onExit(strip(logs))) {
            res.cancel()
          }
        })

        return [
          res.catch(err => {
            if (err.isCanceled) {
              return
            }

            throw err
          }),
          () => createLogsMatcher(strip(logs)),
        ]
      },
    }
  },
}
