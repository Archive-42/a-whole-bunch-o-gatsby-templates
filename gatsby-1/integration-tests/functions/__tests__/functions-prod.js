import { runTests } from "../test-helpers"

runTests(`production`, `http://localhost:9000`)
