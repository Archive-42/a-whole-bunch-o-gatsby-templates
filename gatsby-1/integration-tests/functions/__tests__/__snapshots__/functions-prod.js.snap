// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`production functions can parse different ways of sending data file in multipart/form 1`] = `
Array [
  Object {
    "buffer": Object {
      "data": Array [
        104,
        105,
        32,
        102,
        114,
        105,
        101,
        110,
        100,
        115,
        10,
      ],
      "type": "Buffer",
    },
    "encoding": "7bit",
    "fieldname": "file",
    "mimetype": "text/plain",
    "originalname": "test.txt",
    "size": 11,
  },
]
`;

exports[`production functions can parse different ways of sending data form data 1`] = `
Object {
  "a": "form-data",
}
`;

exports[`production functions can parse different ways of sending data form parameters 1`] = `
Object {
  "a": "form parameters",
}
`;

exports[`production functions can parse different ways of sending data json body 1`] = `
Object {
  "a": "json",
}
`;

exports[`production functions can parse different ways of sending data query string 1`] = `
Object {
  "amIReal": "true",
}
`;

exports[`production functions get parsed cookies cookie 1`] = `
Object {
  "foo": "blue",
}
`;

exports[`production response formats returns json correctly 1`] = `
Object {
  "amIJSON": true,
}
`;

exports[`production response formats returns json correctly 2`] = `
Object {
  "connection": "close",
  "content-length": "16",
  "content-type": "application/json; charset=utf-8",
  "etag": "W/\\"10-R6td1pV+B+Xz9CJkNeaEI2kP+QY\\"",
  "x-powered-by": "Express",
}
`;

exports[`production response formats returns text correctly 1`] = `"I am typescript"`;

exports[`production response formats returns text correctly 2`] = `
Object {
  "connection": "close",
  "content-length": "15",
  "content-type": "text/html; charset=utf-8",
  "etag": "W/\\"f-zwggT56l/fnWBm4dI0PkA4E3i4E\\"",
  "x-powered-by": "Express",
}
`;

exports[`production routing dynamic routes 1`] = `
Object {
  "super": "additional",
  "userId": "23",
}
`;

exports[`production routing dynamic routes 2`] = `
Object {
  "0": "super",
}
`;

exports[`production routing routes with special characters 1`] = `"I-Am-Capitalized.js"`;

exports[`production routing routes with special characters 2`] = `"some whitespace.js"`;

exports[`production routing routes with special characters 3`] = `"with-äöü-umlaut.js"`;

exports[`production routing routes with special characters 4`] = `"some-àè-french.js"`;

exports[`production routing routes with special characters 5`] = `"some-אודות.js"`;

exports[`production routing secondary-level API 1`] = `"I am at a secondary-level"`;

exports[`production routing secondary-level API 2`] = `"I am another sub-directory function"`;

exports[`production routing secondary-level API with index.js 1`] = `"I am an index.js in a sub-directory!"`;

exports[`production routing top-level API 1`] = `"I am at the top-level"`;

exports[`production typescript typescript functions work 1`] = `"I am typescript"`;
